### ISSUE TITLE
Upgrade 2.0

### ISSUE THREAD

**Issue by JohnKhandygo** [l1feforce/FileSplitter#2]
- [ ] Не реализован специальный случай именования выходных файлов.
- [ ] "Нумерация" выходных файлов по умолчанию не соответствует условию.
- [x] В коде не должно быть лишних комментариев. Если для того, чтобы понять Ваш код нужны комментарии, то это плохой код. Особенно, если они не совсем правдивы. 
- [ ] В kotlin вместо контенации строк лучше использовать string interpolation.
- [ ] Файлы в папке input, конечно, нужны для тестирования, но только для него. Такие файлы принято держать в тестовых ресурсах (src/test/resources).
- [ ] Файлы в папке output --- побочный результат запуска тестов, они не должны оставаться в проекте. Для того, чтобы решить эту проблему используйте специальный механизм junit4 [`TemporaryDirectory`](https://junit.org/junit4/javadoc/4.12/org/junit/rules/TemporaryFolder.html) ([пример использования](https://examples.javacodegeeks.com/core-java/junit/junit-temporary-folder-example/)).
- [x] В pom.xml `groupId` и `artifactId` --- это не одно и то же: `groupId` используется как идентификатор **группы** приложений, в то время как `artifactId` --- это **имя** конкретного приложения. 
- [x] У Вас какие-то проблемы с именованием функций/переменных. Постарайтесь дать переменным и методам "говорящие" названия. Для того, чтобы "безболезненно" переименовать любую сущность в idea поставьте курсов на эту сущность (например, название переменной, которую хотите переименовать) и используйте комбинацию клавиш Shift + f6.
- [ ] В тестах не покрыты случаи пустых строк в случае разбиения по строкам и переносов строк в случае разбияния по символам. 
- [x] Не вижу инструкции/скриптов по сборке/запуску приложения из командной строки.
- [ ] Чтение из файла реализовано неэффективно: при больших файлах программа может упасть из-за недостатка выделенной ей памяти из-за того, что все содержимое файла считывается и хранится в памяти программы. То есть, например, для того, чтобы обработать входной файл размером 2GB программе необходимо будет выделить не менее 2GB оперативной памяти для работы (а [`File::readText`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/java.io.-file/read-text.html) в принципе не способен работать с файлами большего объема).
- [ ] Вывод в файл реализован некорректно.

**Comment by l1feforce** [comment]
А как лучше эффективно реализовать чтение файла и корректно сделать вывод в файл?

**Comment by JohnKhandygo** [comment]
Использовать построковое чтение файла через более низкоуровневые операции и обрабатывать строки под одной за раз . То есть прочитали строку, обработали (фактически записали ее или ее часть куда надо), читаете следующую.

**Comment by l1feforce** [comment]
А собрать в консоли через maven это ведь то что нужно?

**Comment by JohnKhandygo** [comment]
Как угодно. Даже не обязательно через консоль. Главное, чтобы была инструкция по сборке (проще всего в readme положить) и можно было через консоль запустить (для этого скрипт, либо пример команды). Ну и настоятельно рекомендую погонять через консоль самостоятельно, иногда какие-то простые ошибки вылезают.

**Comment by l1feforce** [comment]
Частично исправил.

**Comment by l1feforce** [comment]
Можете пожалуйста посмотреть исправления.

**Comment by JohnKhandygo** [comment]
Все, что не отмечено галочками из предыдущих комментариев актуально. 

- [ ] Самое главное: у меня сейчас Ваши тесты падают. В следующий раз я не буду проверять решение, пока тесты не будут исправлены. 
- [ ] В условии задания сказано:
  > При проектировании использовать объектно-ориентированный подход

  У Вас же просто набор статических функций реализован, что к ООП имеет мало отношения.


По предыдущим замечаниям:

> Не реализован специальный случай именования выходных файлов.

Этот случай, конечно, теперь реализован, однако несколько странно. Вместо того, чтобы добавлять три лишних случая к `when` внутри `Parser::main`, можно сначала определить базовое имя выходного файла, а уже затем переходить к разбору ключей. 

> "Нумерация" выходных файлов по умолчанию не соответствует условию.

Там ошибка, Ваши тесты ее не отлавливают.

> Файлы в папке input, конечно, нужны для тестирования, но только для него. Такие файлы принято держать в тестовых ресурсах (src/test/resources).

Во-первых, Вы неправильно разместили папку resources (стандартная структура maven проектов выглядит [так](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html)). Во-вторых, в репозитории почему-то остались старая папка input.

> В тестах не покрыты случаи пустых строк в случае разбиения по строкам и переносов строк в случае разбияния по символам.

Случай пустых строк покрыт, но с чего Вы взяли, что их надо игнорировать?

**Comment by l1feforce** [comment]
Как необходимо нумеровать выходные файлы если их число больше чем 2 английских алфавита? То есть что необходимо после  filenameba, filenamebb, ..., filenamezz ?

**Comment by JohnKhandygo** [comment]
Давайте считать, что это должно приводить к ошибке.

**Comment by l1feforce** [comment]
> можно сначала определить базовое имя выходного файла, а уже затем переходить к разбору ключей

Не совсем понял как это.

**Comment by JohnKhandygo** [comment]
Можно записать значение аргумента `outputFilesName`  (для вызова метода `splitByX`) в переменную, а потом делать `when`.

**Comment by l1feforce** [comment]
А подскажите пожалуйста, можно ли будет после окончания 2 задания выбрать третьим заданием приложение на андроид? И если да, то какой сложности будет достаточно?

**Comment by JohnKhandygo** [comment]
Да, можно. Возможно, в этом случае Вас будет проверять другой преподаватель. Сложность зависит от того, какого рода приложение Вы собираетесь писать. Если просто standalone приложение, то уровня шашек для двух пользователей на одном устройстве.

**Comment by l1feforce** [comment]
Решил писать на kotlin с использованием библиотеки Anko, а можно ли кому то задать вопросы именно по работе с библиотекой?

**Comment by JohnKhandygo** [comment]
Не знаю. Вы же сами выбрали эту библиотеку.

**Comment by l1feforce** [comment]
Черновик 3 задания: https://github.com/l1feforce/BaldaGame

**Comment by JohnKhandygo** [comment]
1. Второе задание еще не зачтено. Вы исправили все замечания? Задание можно повторно проверять?
2. Вы не согласовали условие второго задания. Необходимо составить описание правил и функциональности приложения, которую Вы обязуетесь реализовать, в противном случае не понятно, как выставлять оценку. Описание следует разместить в issue в репозитории с реализацией.

**Comment by JohnKhandygo** [comment]
Из того, что я вижу прямо сейчас: в проекте недостаточно тестов для того, чтобы можно было сделать вывод о том, насколько корректно функционирует приложение. Ну и не факт, что кто-то из других преподавателей сейчас будет готов взяться за проверку Вашего задания.

**Comment by l1feforce** [comment]
На данный момент во второй работе не сделан переход к ООП, нет реализации временных директорий, и не исправлены реализации чтения из файла и записи в файл. По второму пункту не совсем понял, есть вопросы по реализации, но вроде весь необходимый функционал есть? Какие критерии выставления оценки? Или речь про третье задание?

**Comment by l1feforce** [comment]
https://github.com/l1feforce/BaldaGame
Cсылка на третье задание, игра уже рабочая, список того что еще хочу реализовать есть в ReadMe.
Что необходимо для допуска к экзамену?

**Comment by l1feforce** [comment]
Разместил описание и видео примера работы игры в issue и readme.
